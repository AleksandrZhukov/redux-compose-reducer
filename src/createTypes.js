import {
  namespacedActionType,
  createTypesObject,
  createCapturedError,
} from './utils'

const createTypes = (namespace, types = []) => {
  if (
    typeof namespace !== 'string' ||
    namespace.length === 0 ||
    !Array.isArray(types) ||
    types.length === 0
  )
    throw createError(ARGUMENT_ERROR)

  const result = createTypesObject()
  types.forEach(type => {
    result[type] = namespacedActionType(namespace, type)
  })
  return result
}

const createError = message => createCapturedError(message, createTypes)

export const ARGUMENT_ERROR =
  'Expected first argument (namespace) to be a non empty string ' +
  'and second argument (types) to be a non empty array.'

export default createTypes
