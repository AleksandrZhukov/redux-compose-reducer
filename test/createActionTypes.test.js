import createTypes, { ARGUMENT_ERROR } from '../src/createTypes'

describe('createTypes', () => {
  it('produces correct action types', () => {
    const actual = createTypes('test', ['action1', 'action2'])
    expect(Object.keys(actual)).toEqual(['action1', 'action2'])
    expect(Object.values(actual)).toEqual(['test/action1', 'test/action2'])
  })

  describe('proxy', () => {
    describe('enviroments without Proxy', () => {
      const proxy = global.Proxy

      beforeAll(() => {
        global.Proxy = undefined
      })

      afterAll(() => {
        global.Proxy = proxy
      })

      it('does not produce error', () => {
        const actual = createTypes('test', ['action1', 'action2'])
        expect(() => actual.missingType).not.toThrowError()
      })
    })

    describe('access to action that does not exists', () => {
      let captureStackTrace

      beforeAll(() => {
        captureStackTrace = Error.captureStackTrace
      })

      afterAll(() => {
        Error.captureStackTrace = captureStackTrace
      })

      it('produces error', () => {
        Error.captureStackTrace = jest.fn()

        const actual = createTypes('test', ['action1', 'action2'])
        expect(() => actual.missingType).toThrowErrorMatchingSnapshot()

        expect(Error.captureStackTrace.mock.calls.length).toBe(1)
      })

      it('works in enviroments without captureStackTrace', () => {
        Error.captureStackTrace = undefined
        const actual = createTypes('test', ['action1', 'action2'])
        expect(() => actual.missingType).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe('validation', () => {
    describe('without parameters', () => {
      it('produces error', () => {
        expect(() => createTypes()).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('with incorrect namespace', () => {
      it('produces error', () => {
        expect(() => createTypes({})).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('without types', () => {
      it('produces error', () => {
        expect(() => createTypes('test')).toThrowError(ARGUMENT_ERROR)
      })
    })

    describe('with incorrect types', () => {
      it('produces error', () => {
        expect(() => createTypes('test', {})).toThrowError(ARGUMENT_ERROR)
      })
    })
  })
})
