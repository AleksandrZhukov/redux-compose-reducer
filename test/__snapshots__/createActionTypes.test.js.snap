// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTypes proxy access to action that does not exists produces error 1`] = `"Key 'missingType' does not exist."`;

exports[`createTypes proxy access to action that does not exists works in enviroments without captureStackTrace 1`] = `"Key 'missingType' does not exist."`;
